### ISSUE TITLE
State of inline style commands

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/editor-recommendations#32]
Problem: What should be the state (on/off) of an inline style (e.g. bold) command depending on a selection? In other words – when is the feature active (e.g. its button pressed)?

There are multiple concerns here:
- what if a selection starts and ends in different elements (e.g. `<strong>fo[o</strong>b]ar`)?
- how the selection directions affects this (selection has a focus and anchor)?
- how to handle multi-range selections?
- what is the relation between the command state and how backspace/delete/typing works?

**Comment by Reinmar** [comment]
We've made some tests here how the browsers handle it by default. We tested Chrome, FF and IE11. Chrome and IE11 were consistent here in implementing the gravity (working to the left). FF has some more sophisticated implementation, but it also handles navigation with arrow keys in a more sophisticated way.

## Gravity

Gravity makes the "input" stick more to the left or to the right. It's popular that gravity is set to left, so when you type at the end of a bolded word the new letters are bolded (inherit styles from the left of the caret).

It seems that Chrome and IE uses gravity for backspace and typing. E.g. regardless of how the selection has been made (from L to R or R to L), pressing backspace in this case: `<strong>fo[o</strong>b]ar` puts the selection inside the `<strong>` tag. If you typed instead of just pressing backspace (typing on a non-empty selection is handled as delete+typing), then the letter will appear bolded.

The opposite will happen when you type/delete on the left side of a bolded text. The caret will land outside of `<strong>` and the text will not be bolded. That's because it inherited the styles from the left side.

## Idea

In my opinion the state of the command should reflect the style that a letter inserted into such a selection would have. So if the left side of your selection (don't confuse this with anchor and focus which are irrelevant here) is outside bold, then the bold command is disabled. If inside, then it is enabled.

**Comment by fredck** [comment]
+1 for left-gravity on the reading direction start of the selection.

**Comment by fredck** [comment]
> FF has some more sophisticated implementation, but it also handles navigation with arrow keys in a more sophisticated way.

I always liked this features. I call it "**stay inside/outside**". The idea is that if you navigate with the arrows to the boundary of an inline element, if you were inside the element, you stay inside. If you were outside, you say outside.

This feature is nice because it give a bit of control for the user. I'm all for it, but it can stay optional as it is not essential and may be tricky to implement.

Still the default left-gravity applies for mouse-click.

**Comment by fredck** [comment]
> what is the relation between the command state and how backspace/delete/typing works?

No doubt that it should always reflect the command state. 

For non collapsed selection, these operations should be interpreted as "delete the selection contents, keeping the beginning of the now collapsed selection at the same position". In the case of backspace and delete, nothing else happens. Typing of course will insert the character at the new selection position.

**Comment by fredck** [comment]
> how the selection directions affects this (selection has a focus and anchor)?

I don't think it should make any difference. It would be confusing for users. Better to see this as "beginning/end of the selection in the reading direction".

**Comment by Reinmar** [comment]
> I always liked this features. I call it "stay inside/outside". The idea is that if you navigate with the arrows to the boundary of an inline element, if you were inside the element, you stay inside. If you were outside, you say outside.

I like it as well. And it concerns only collapsed selection AFAIK, so its relation with the state command is clear. So (fortunately, as it simplifies many things) it's only keyboard handling feature.

**Comment by scofalik** [comment]
The state (attributes) of selection thing is more complicated than we initially thought. We cannot look always to the left-side of selection beginning. Consider following:

I have text "abcfooxyz". I select "foo" and apply bold. The text is now "abc[**foo**]xyz". The whole text in selection is bolded. But if we will look at the left-side, "c" is not bold so the selection wouldn't have bold attribute, meaning that the bold command would be off.

Another example: We have a list item, which whole content is bold:
- ^**foobar**

When we start typing at `^` should we type in bold?

Another example: We have an image and a character after the image and collapsed selection between them. Selection should have attributes of character not the image, for sure.

So here is what we figured out:
1. If selection is not-collapsed, look at the first character in the selected range and take it's attributes.
2. If selection is collapsed or there were no characters, look at the node before caret position. If it's a character, take it's attributes.
3. If not, look at the node after caret position. If it's a character, take it's attributes.
4. If not, take attributes of parent element.

This should cover all bases. Please review this algorithm.

**Comment by scofalik** [comment]
After putting some thought into how applying command in empty node should work I am having some doubts.

Currently:
- when node is empty or selection is between two non-character nodes (most often images), we get attributes from selection's parent node ("default" style),
- when node is empty, we set attributes on selection's parent node (set "default" style).

Though these are edge cases, it may be a little confusing, that applying style in one place of node has an effect on entire node. Moreover, it is very tricky (almost impossible) to change the "default" style - you would have to remove whole content of a paragraph. But, frankly, "default" styles are applied only in very specific cases (empty node or peculiar scenario of selection between two non-characters).

Having attributes on parent's node is useful probably only in one case: when we split the node (use enter key, create new paragraph / list item / etc.). Then, we want to type in new paragraph with the same styles as before hitting enter. This could be done simply by storing selection attributes, splitting node and re-storing selection attributes. Only drawback is if we do not type anything, change selection and then get back - our styles will be lost.

Even if we want to keep the styles in new node, we can't simply copy attributes of the node that has been split, as the styles of the last characters in split node might be different than attributes of the split node. This would bring even more confusion, so we have to take selection's attributes and apply them to newly created node.

Do you have any other use cases for having inline-style attributes (bold, etc.) on parent nodes? It is not crazy to abandon this idea.

**Comment by fredck** [comment]
I agree that no inline styles should endup in the parent. They can be there just as a way to set the selection's default when the parent is empty.

**Comment by scofalik** [comment]
After F2F discussion with @fredck we decided that, at least for now, this will be expected behavior:
- type in paragraph, change styles (i.e. apply bold),
- click enter key, empty paragraph is created,
- selection will keep styles applied just before clicking enter,
- move caret and go back to the empty paragraph,
- selection won't have any styles (so styles applied before clicking enter are not saved).

This might not be the best solution from UX point of view but it has its merits (mostly it lets us keep things simpler, and any other solution we came up with was introducing some confusing edge cases).

Still, we should have our minds open on that matter and possibly will review this behavior in future.

**Comment by scofalik** [comment]
Also we've updated an algorithm that sets selection style:
1. If selection is not-collapsed, look at the first character in the selected range and take it's attributes.
2. If selection is collapsed or there were no characters, look at the node before caret position. If it's a character, take it's attributes.
3. If not, look at the node after caret position. If it's a character, take it's attributes.
4. If not, try to find a first character that is on left of caret and if found, take it's attributes.
5. If not found, try to find a first character that is on the right of caret and if found, take it's attributes.
6. If not found, selection has no attributes (styles).

**Comment by wwalc** [comment]
Just adding a cross link to an issue for CKEditor 4: https://dev.ckeditor.com/ticket/12634 because apparently we are about to replicate the same behavior (considered as invalid by some users).
Note point 7 there:

> Check the DOM - strong elements are still there.

**Comment by Reinmar** [comment]
https://dev.ckeditor.com/ticket/12634 is a very popular issue – it has a lot of DUPs (much more than we had noticed, because it can be reported in many various ways). Let's not repeat the mistakes of browser vendors here. Users expect that styles will be maintained in empty paragraphs and popular editors handle this as expected (I checked MS Word, GDocs and TinyMCE which workarounds the browser behavior).

So there are three use cases which IMO we must handle correctly:
- returning to an empty paragraph by arrow keys,
- returning to an empty paragraph by mouse click,
- returning to an empty paragraph by pressing backspace.

In all of them the styles that the text would have there initially, should be maintained. It's editor's implementator decision how to do this. For CKEditor 5 we'll continue the discussion in https://github.com/ckeditor/ckeditor5-design/issues/128.

**Comment by scofalik** [comment]
I've pushed branch `t/32` https://github.com/ckeditor/editor-recommendations/tree/t/32 with initial draft for this usability.

**Comment by Comandeer** [comment]
I've merged that branch with `gh-pages` and draft is publicly accessible under http://ckeditor.github.io/editor-recommendations/usability/inline-style-state.html.

**Comment by fredck** [comment]
Please fix the case style of the headers in the text to "Match What We Have In Other Pages" :)

**Comment by Comandeer** [comment]
Actually we don't have… That Type Of Headers Are Only On Homepage And About ones.

I'll create new issue for it.

**Comment by fredck** [comment]
> Actually we don't have… That Type Of Headers Are Only On Homepage And About ones.

... And Features And Use Cases And... :)

**Comment by Comandeer** [comment]
The draft is hanging without changes for over a year. Are we going to withdraw it or enhance it and make a recommendation?
